### PR TITLE
INTMDB-655: PAK Resource Updates + Doc Cleanup + Deprecation Warnings

### PR DESCRIPTION
**Pull request by martinstibbe**
## Description

PAK Resource Updates + Doc Cleanup + Deprecation Warnings

Link to any related issue(s):

## Type of change:

- [ ] Bug fix (non-breaking change which fixes an issue). Please, add the "bug" label to the PR.
- [ ] New feature (non-breaking change which adds functionality). Please, add the "enhancement" label to the PR.
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected). Please, add the "breaking change" label to the PR.
- [x] This change requires a documentation update
- [x] Documentation fix/enhancement

## Required Checklist:

- [x] I have signed the [MongoDB CLA](https://www.mongodb.com/legal/contributor-agreement)
- [x] I have read the [contribution guidelines](https://github.com/mongodb/terraform-provider-mongodbatlas/blob/master/CONTRIBUTING.md)
- [x] I have added tests that prove my fix is effective or that my feature works per HashiCorp requirements
- [x] I have added any necessary documentation (if appropriate)
- [x] I have run make fmt and formatted my code

### CODE REVIEW

File: mongodbatlas/resource_mongodbatlas_project.go
Comment (andreaangiolillo) [inline mongodbatlas/resource_mongodbatlas_project.go]: Could you add this message to the `provider.go` (similar to  [provider.go#L42](https://github.com/mongodb/terraform-provider-mongodbatlas/blob/b4045e1c5f98432a1127495e8ef238fd092ab513/mongodbatlas/provider.go#L42))?
It would be great to have only one deprecation message that we can use in all the resources/data sources. Thanks!

File: mongodbatlas/resource_mongodbatlas_project_api_key.go
Comment (andreaangiolillo) [inline mongodbatlas/resource_mongodbatlas_project_api_key.go]: this piece of code seems to be duplicated in other parts of this file. Could you extract it in a function?

Comment (martinstibbe) [reply mongodbatlas/resource_mongodbatlas_project_api_key.go]: Added function to simplify

File: mongodbatlas/resource_mongodbatlas_project.go
Comment (martinstibbe) [reply mongodbatlas/resource_mongodbatlas_project.go]: Added deprecation message from provider.go

File: mongodbatlas/resource_mongodbatlas_project.go
Comment (andreaangiolillo) [inline mongodbatlas/resource_mongodbatlas_project.go]: this deprecation message does not seem correct. Here we are deprecating a param but the resource is not deprecated. I would expect to see another deprecation message define in provider.go for deprecating a parameter. Am I missing something here?

Comment (martinstibbe) [reply mongodbatlas/resource_mongodbatlas_project.go]: ok will make one specific to this unique case as it is not parameter to parameter change it is parameter to different resource

Comment (andreaangiolillo) [reply mongodbatlas/resource_mongodbatlas_project.go]: The main reason to have `"this parameter is deprecated, please use project API Key resource instead"` defined in `provider.go` as a constant is to use the same message once we have another parameter to deprecate in another resource. Feel free to ignore this comment if you think it is not necessary, we can always address this later

Comment (martinstibbe) [reply mongodbatlas/resource_mongodbatlas_project.go]: Added new message

**Review by andreaangiolillo — APPROVED** [review]
LGTM. Thanks for all the changes

File: website/docs/r/access_list_api_key.html.markdown
Comment (themantissa) [inline website/docs/r/access_list_api_key.html.markdown]: "Atlas resources require that all API requests originate from IP addresses on the API access list." is not actually true as written.  Access Lists are super confusing with the Admin API.  Basically 1) some endpoints require an access list (like backup) while others do not (like clusters) by **default**.  2) Users can select to require an API key access list in the org settings in the UI with the "Require IP Access List for the Atlas Administration API" setting.  So we should make this confusing nuance clear (Last I heard IAM was going to clean this up so @Zuhairahmed we might want to check in with Fuat on that)

File: examples/atlas-api-key/create-and-assign-pak-seperately/README.md
Comment (themantissa) [inline examples/atlas-api-key/create-and-assign-pak-seperately/README.md]: All three examples are really good but this whole area can be hard to understand if the UI is your main PAK experience.  In the UI one has organization keys and project keys - if one looks closely you can tell  all are really  org keys and are simply assigned to projects but that's not as clear as it is in the Admin API.  I think an intro README at the atlas-api-key folder level explaining this and why we provided 3 examples would be very useful.   It's not a blocker to this PR but I think would be a good follow up, perhaps for the docs team?  cc @Zuhairahmed

File: examples/atlas-api-key/create-and-assign-pak-seperately/main.tf
Comment (themantissa) [inline examples/atlas-api-key/create-and-assign-pak-seperately/main.tf:9]: Should we provide a warning that this way is being deprecated in the near future?

File: mongodbatlas/resource_mongodbatlas_project_api_key.go
Comment (themantissa) [inline mongodbatlas/resource_mongodbatlas_project_api_key.go:236]: Should we add a test to cover the changed functionality?

File: website/docs/d/access_list_api_key.html.markdown
Comment (themantissa) [inline website/docs/d/access_list_api_key.html.markdown]: See comment for resource about access lists and the Admin API.

File: website/docs/r/project_api_key.html.markdown
Comment (themantissa) [inline website/docs/r/project_api_key.html.markdown:9]: Do we want to add an example for project_assignment?

File: website/docs/r/project_api_key.html.markdown
Comment (themantissa) [inline website/docs/r/project_api_key.html.markdown]: Add project_assignment?

File: mongodbatlas/resource_mongodbatlas_project.go
Comment (themantissa) [inline mongodbatlas/resource_mongodbatlas_project.go:85]: The docs for project should also be updated with a deprecation message.

**Review by themantissa — CHANGES_REQUESTED** [review]
Comments inline, thank you!

**Comment by themantissa** [comment]
@Zuhairahmed are we going to have our docs team work on this upgrade guide part of the scope?

File: examples/atlas-api-key/create-api-key-assign-to-multiple-projects/variables.tf
Comment (zach-carr) [inline examples/atlas-api-key/create-api-key-assign-to-multiple-projects/variables.tf]: Looks like it should be "id"
```suggestion
  description = "Atlas project id"
```

File: examples/atlas-api-key/create-api-key-assign-to-multiple-projects/variables.tf
Comment (zach-carr) [inline examples/atlas-api-key/create-api-key-assign-to-multiple-projects/variables.tf]: Description needs an update
```suggestion
variable "additional_project_id" {
  description = "Atlas project id"
```

File: mongodbatlas/resource_mongodbatlas_project_api_key.go
Comment (zach-carr) [inline mongodbatlas/resource_mongodbatlas_project_api_key.go]: Super minor, but just for consistency in the comments, these could all start with capitalized verbs
```suggestion
		// Getting the current api_keys and the new api_keys with changes
```

File: mongodbatlas/resource_mongodbatlas_project_api_key.go
Comment (zach-carr) [inline mongodbatlas/resource_mongodbatlas_project_api_key.go]:
```suggestion
		// Adding new api_keys into the project
```

File: website/docs/d/access_list_api_key.html.markdown
Comment (zach-carr) [inline website/docs/d/access_list_api_key.html.markdown]: I think adding a comma to "...cidr_block  or ip_address`,` not both" makes this a little clearer, but I suggest rephrasing to get rid of the colon:
```suggestion
-> **NOTE:** You must set either the `cidr_block` attribute or the `ip_address` attribute. Don't set both.
```

File: website/docs/d/access_list_api_keys.html.markdown
Comment (zach-carr) [inline website/docs/d/access_list_api_keys.html.markdown:67]: The comment above applies here, too

File: website/docs/r/access_list_api_key.html.markdown
Comment (zach-carr) [inline website/docs/r/access_list_api_key.html.markdown:48]: The comment about this note applies here

File: examples/atlas-api-key/create-and-assign-pak-seperately/README.md
Comment (zach-carr) [inline examples/atlas-api-key/create-and-assign-pak-seperately/README.md]: [typo]
```suggestion
# MongoDB Atlas Provider -- Create and Assign PAK Separately
```

File: examples/atlas-api-key/create-and-assign-pak-seperately/README.md
Comment (zach-carr) [inline examples/atlas-api-key/create-and-assign-pak-seperately/README.md]: Small suggestion for clarity
```suggestion
This example creates an organization API Key and access list, showing how to attach a CIDR block.
```

File: examples/atlas-api-key/create-and-assign-pak-seperately/README.md
Comment (zach-carr) [inline examples/atlas-api-key/create-and-assign-pak-seperately/README.md]: Editing for clarity and consistency 
```suggestion
In this example, we will set up an organization API key and attach an access list to it.
```

File: examples/atlas-api-key/create-and-assign-pak-together/README.md
Comment (zach-carr) [inline examples/atlas-api-key/create-and-assign-pak-together/README.md]: Suggestion copied from above
```suggestion
This example creates a project API Key and access list, showing how to attach a CIDR block.

Variables Required to be set:
- `project_id`: ID of the Atlas project
- `org_id`: ID of Atlas organization
- `public_key`: Atlas public key
- `private_key`: Atlas  private key

In this example, we will set up a project API key and attach an access list to it.
```

File: examples/atlas-api-key/create-api-key-assign-to-multiple-projects/README.md
Comment (zach-carr) [inline examples/atlas-api-key/create-api-key-assign-to-multiple-projects/README.md]: Similar suggestions
```suggestion
This example creates a project API key and assigns it to multiple projects.

Variables Required to be set:
- `project_id`: ID of the Atlas project that created the API key
- `public_key`: Atlas public key
- `private_key`: Atlas  private key
- `additional_project_id`: Additional project ID, used to assign the API key to more than one project at a time

In this example, we will set up a project API key and attach it to multiple projects.
```

**Review by zach-carr — CHANGES_REQUESTED** [review]
Lots of very minor suggestions (sorry 😅), but I think all together they'll improve the flow of the page

File: website/docs/r/access_list_api_key.html.markdown
Comment (Zuhairahmed) [reply website/docs/r/access_list_api_key.html.markdown]: @themantissa good news, feedback from Fuat is that The restriction about #1 will be removed in a month or so, which means we'll have no IP access list requirement by default on any resource. Created [DOCSP-30474](https://jira.mongodb.org/browse/DOCSP-30474) to flow these new updates into all impacted resources and data sources in our Terraform Docs as well. Suggest for now we can commit your suggestion and then when Atlas Admin API has been updated we can then update all our Terraform docs for this fix at that time together.

File: examples/atlas-api-key/create-and-assign-pak-seperately/main.tf
Comment (Zuhairahmed) [reply examples/atlas-api-key/create-and-assign-pak-seperately/main.tf:9]: +1 to deprecation comment, would also specify that they are going away in v1.12.0 release (2 versions out)

File: website/docs/r/project_api_key.html.markdown
Comment (martinstibbe) [reply website/docs/r/project_api_key.html.markdown:9]: Added create-api-key-assign-to-multiple-projects

File: website/docs/r/project_api_key.html.markdown
Comment (martinstibbe) [reply website/docs/r/project_api_key.html.markdown]: Added

File: mongodbatlas/resource_mongodbatlas_project.go
Comment (martinstibbe) [reply mongodbatlas/resource_mongodbatlas_project.go:85]: Added

File: examples/atlas-api-key/create-and-assign-pak-seperately/README.md
Comment (Zuhairahmed) [inline examples/atlas-api-key/create-and-assign-pak-seperately/README.md]:
```suggestion
# MongoDB Atlas Provider -- Create and Assign PAK Separately (Deprecated) 

Note this functionality has been deprecated and will be removed in v1.12.0. Instead suggest see examples for Create and Assign PAK Together + Create and Assign PAK to Multiple Projects.
```

File: mongodbatlas/resource_mongodbatlas_project_api_key.go
Comment (Zuhairahmed) [reply mongodbatlas/resource_mongodbatlas_project_api_key.go:236]: just confirming that @martinstibbe committed a new _test.go file which should address this comment

File: examples/atlas-api-key/create-and-assign-pak-seperately/README.md
Comment (Zuhairahmed) [reply examples/atlas-api-key/create-and-assign-pak-seperately/README.md]: Added this atlas-api-key README file directly, feel free to share any additional comments @zach-carr and @themantissa!

File: examples/atlas-api-key/README.md
Comment (Zuhairahmed) [inline examples/atlas-api-key/README.md:1]: Added this atlas-api-key README file directly (per below feedback), feel free to share any additional comments @zach-carr and @themantissa!

**Comment by Zuhairahmed** [comment]
> @Zuhairahmed are we going to have our docs team work on this upgrade guide part of the scope?

Added additional commit to this PR with PAK Upgrade Guide to address [above comment](https://github.com/mongodb/terraform-provider-mongodbatlas/pull/1208#issuecomment-1579419346). Feel free to review/layer in any additional comments @themantissa @zach-carr. 

Also separately we will have our usual v1.10.0 upgrade guide as well (will start on this after pre-release is out) which will include all updates as part of v1.10.0 release and will link to this dedicated PAK Upgrade Guide one.

File: website/docs/r/project.html.markdown
Comment (Zuhairahmed) [inline website/docs/r/project.html.markdown]:
```suggestion
**WARNING:** The `api_keys` parameter is deprecated and will be removed in v1.12.0 release from codebase. Use `mongodbatlas_project_api_key`  resource instead. 
```

**Review by Zuhairahmed — APPROVED** [review]
LGTM with a few minor comments. also added a few commits directly as well

File: examples/atlas-api-key/README.md
Comment (zach-carr) [inline examples/atlas-api-key/README.md]: suggestion, feel free to take or leave - I think this first sentence could be simplified to something like:
"The most notable enhancement among these was the deprecation of the `api_keys` parameter from the ..."

otherwise, just a tweak to what's there:
```suggestion
The most notable improvement among these as part of v1.10.0 release was the deprecation of the `api_keys` parameter from the `mongodbatlas_project` resource. We also extended the functionality of the `mongodbatlas_project_api_key` resource by incorporating the `project_assignment` parameter. This enhancement removes the necessity for users to create multiple `mongodbatlas_project` resource blocks just to assign keys. This update streamlines the process of assigning an API Key to multiple projects, making it less cumbersome and more manageable.
```

**Review by zach-carr — APPROVED** [review]
LGTM

**Review by andreaangiolillo — APPROVED** [review]
LGTM

File: examples/atlas-api-key/README.md
Comment (themantissa) [reply examples/atlas-api-key/README.md:1]: Not blocking but I think it would be helpful to explain the org/project API keys as I think that can still be confusing when using the API vs UI - i.e. All PAKs are organization API keys (org member role) but can be assigned to 1 or more projects in the org (such as for project owner role).

File: mongodbatlas/resource_mongodbatlas_project_api_key.go
Comment (themantissa) [reply mongodbatlas/resource_mongodbatlas_project_api_key.go:236]: TY!

File: website/docs/r/access_list_api_key.html.markdown
Comment (themantissa) [reply website/docs/r/access_list_api_key.html.markdown]:
```suggestion
    Creates the access list entries for the specified Atlas Organization API Key. 
```

Comment (themantissa) [reply website/docs/r/access_list_api_key.html.markdown]: @Zuhairahmed that's great news!  My suggestion then is just to remove this caveat for now since it's not technically correct.   Writing out what is correct is a bit of work so we might want to do that as a follow up.

File: examples/atlas-api-key/create-and-assign-pak-seperately/main.tf
Comment (themantissa) [reply examples/atlas-api-key/create-and-assign-pak-seperately/main.tf:9]: Not changed but not blocking - would be good to add.

File: website/docs/d/access_list_api_key.html.markdown
Comment (themantissa) [reply website/docs/d/access_list_api_key.html.markdown]:
```suggestion
    Displays the access list entries for the specified Atlas Organization API Key. 
```
Making same suggestion as above.  Not blocking but nice to correct.

File: website/docs/r/project_api_key.html.markdown
Comment (themantissa) [reply website/docs/r/project_api_key.html.markdown:9]: Thank you!

File: website/docs/r/project_api_key.html.markdown
Comment (themantissa) [reply website/docs/r/project_api_key.html.markdown]: TY!

File: website/docs/r/project.html.markdown
Comment (themantissa) [reply website/docs/r/project.html.markdown]: @Zuhairahmed this will need updating on the project data sources as well.   Maybe do as a follow up?   I'm going to approve this to unblock.

File: mongodbatlas/resource_mongodbatlas_project.go
Comment (themantissa) [reply mongodbatlas/resource_mongodbatlas_project.go:85]: Thank you!

**Review by themantissa — APPROVED** [review]
Nothing blocking this but a few of docs improvements we can make now or do next round (cc @Zuhairahmed).  Thank you for all the changes!

**Review by Zuhairahmed — CHANGES_REQUESTED** [review]
@martinstibbe can you flow through all above doc edits? it's mostly committing most of edits suggested by Zach, Melissa, and myself above and few new ones. after that I can do one last check and then merge. thanks!

**Comment by martinstibbe** [comment]
> @martinstibbe can you flow through all above doc edits? it's mostly committing most of edits suggested by Zach, Melissa, and myself above and few new ones. after that I can do one last check and then merge. thanks!

completed suggested changes

File: examples/atlas-api-key/create-and-assign-pak-seperately/main.tf
Comment (Zuhairahmed) [reply examples/atlas-api-key/create-and-assign-pak-seperately/main.tf:9]: added!

File: examples/atlas-api-key/README.md
Comment (Zuhairahmed) [reply examples/atlas-api-key/README.md:1]: added!

**Review by Zuhairahmed — APPROVED** [review]
LGTM and confirming @themantissa that believe we captured all requested doc improvements as part of this PR

**Review by themantissa — APPROVED** [review]
Thank you for the fixes!  LGTM!

**Pull request merged**
